client/thin-replica-client/src/grpc_connection.cpp

```diff
@@ -356,27 +374,148 @@ TrsConnection::Result TrsConnection::readHash(Hash* hash) {
   return Result::kFailure;
 }
 
-TrsConnection::Result TrsConnection::openStateSnapshotStream(
-    const vmware::concord::replicastatesnapshot::StreamSnapshotRequest& request) {
-  // TODO: Add implementation
-  ConcordAssert("openStateSnapshotStream should not be called. It is unimplemented." && false);
+GrpcConnection::Result GrpcConnection::openStateSnapshotStream(
+    const vmware::concord::replicastatesnapshot::StreamSnapshotRequest& request, RequestId& request_id) {
+  ConcordAssertNE(rss_stub_, nullptr);
+  request_id = current_req_id_.fetch_add(1u);
+
+  vmware::concord::replicastatesnapshot::StreamSnapshotRequest req;
+  req.set_snapshot_id(request.snapshot_id());
+  req.set_last_received_key(request.last_received_key());
+
+  std::unique_ptr<grpc::ClientContext> snapshot_context = std::make_unique<grpc::ClientContext>();
+  snapshot_context->AddMetadata("client_id", client_id_);
+
+  std::unique_ptr<grpc::ClientReaderInterface<vmware::concord::replicastatesnapshot::StreamSnapshotResponse>>
+      snapshot_stream;
+
+  auto stream = async(launch::async, [this, &req, &snapshot_context] {
+    return rss_stub_->StreamSnapshot(snapshot_context.get(), req);
+  });
+  auto status = stream.wait_for(data_timeout_);
+  if (status == future_status::timeout || status == future_status::deferred) {
+    snapshot_context->TryCancel();
+    stream.wait();
+    snapshot_context.reset();
+
+    // If StreamSnapshot did end up returning a pointer to an allocated
+    // stream, make sure it does not get leaked.
+    snapshot_stream = stream.get();
+    snapshot_stream.reset();
+    current_req_id_.fetch_sub(1u);
+    request_id = 0u;
+    return Result::kTimeout;
+  }
+
+  ConcordAssert(status == future_status::ready);
+  snapshot_stream = stream.get();
+  if (snapshot_stream) {
+    WriteLock write_lock(rss_streams_mutex_);
+    rss_streams_.emplace(request_id, std::make_pair(std::move(snapshot_context), std::move(snapshot_stream)));
+    return Result::kSuccess;
+  }
+
+  current_req_id_.fetch_sub(1u);
+  request_id = 0u;
+  snapshot_stream.reset();
+  snapshot_context.reset();
   return Result::kFailure;
 }
 
-void TrsConnection::cancelStateSnapshotStream() {
-  // TODO: Add implementation
-  ConcordAssert("cancelStateSnapshotStream should not be called. It is unimplemented." && false);
-  return;
+void GrpcConnection::cancelStateSnapshotStream(RequestId request_id) {
+  {
+    ReadLock read_lock(rss_streams_mutex_);
+    auto it = rss_streams_.find(request_id);
+    if (it != rss_streams_.end()) {
+      if (!(it->second).second) {
+        ConcordAssertEQ((it->second).first, nullptr);
+        return;
+      }
+      ConcordAssertNE((it->second).first, nullptr);
+      (it->second).first->TryCancel();
+      (it->second).first.reset();
+      (it->second).second.reset();
+    }
+  }
+  {
+    WriteLock write_lock(rss_streams_mutex_);
+    rss_streams_.erase(request_id);
+  }
+}
+
+void GrpcConnection::cancelStateSnapshotStream() {
```

Comment:
You are cancelling all snapshot streams. I recommend to reflect that in the function name via `all` or using `Stream*s*`

Reply:
thanks @robem, very good comments, let me change this.